pkg/utils/symbol_table.go

```diff
@@ -0,0 +1,212 @@
+package utils
+
+import (
+	"errors"
+	"sort"
+	"sync"
+)
+
+// The Symbol interface defines what is needed from a symbol implementation in
+// order to facilitate the lookup functionalities provided by SymbolTable.
+// Implementations of Symbol can hold various types of information relevant to
+// the type of symbol they represent.
+type Symbol interface {
+	Name() string
+	Address() uint64
+	Contains(address uint64) bool
+}
+
+// SymbolTable is used to hold information about symbols (mapping from symbolic
+// names used in code to their address) in a certain executable.
+// It can be used to hold symbols from an ELF binary, or symbols of the entire
+// kernel and its modules.
+// It provides functions to lookup symbols by address and name.
+type SymbolTable[T Symbol] struct {
+	mu sync.RWMutex
+	// All symbols sorted by their address in descending order,
+	// for quick binary searches by address.
+	sortedSymbols []*T
+	// If lazyNameLookup is true, the symbolsByName map
+	// will be populated only when a failed lookup occurs.
+	symbolsByName  map[string][]*T
+	lazyNameLookup bool
+}
+
+var ErrSymbolNotFound = errors.New("symbol not found")
+
+// Creates a new SymbolTable. If lazyNameLookup is true, the mapping from
+// name to symbol will be populated only when a failed lookup occurs.
+// This reduces memory footprint at the cost of the time it takes to lookup
+// a symbol name for the first time.
+func NewSymbolTable[T Symbol](lazyNameLookup bool) *SymbolTable[T] {
+	return &SymbolTable[T]{
+		sortedSymbols:  make([]*T, 0),
+		symbolsByName:  make(map[string][]*T),
+		lazyNameLookup: lazyNameLookup,
+	}
+}
+
+// Adds a slice of symbols to the symbol table.
+func (st *SymbolTable[T]) AddSymbols(symbols []*T) {
+	st.mu.Lock()
+	defer st.mu.Unlock()
+
+	// Add the new symbols to the sorted slice (which now becomes unsorted).
+	// Allocate the slice with the needed capacity to avoid overallocation.
+	oldSymbols := st.sortedSymbols
```

Comment:
Pretty sure this "copy" is unsafe as its just a pointer reassignment. 
I think you should rather allocate the new slice, use on copy the old and parameter slice into the new allocation, then assign that one to the object field (while the prior slice will be given to GC).

Reply:
It's not a copy, I'm just saving a reference to the old slice in a temporary variable before assigning an empty slice to the struct variable.

Reply:
The overall operation you're doing is copying the old into a new extended. I was pinning to the particular line here as its:
1. The first of the operation
2. Seemed problematic to me

Reply:
I still don't understand why it's unsafe

Reply:
You assign the _pointer_ of `st.sortedSymbols` to `oldSymbols`. This isn't a copy.
Then to the pointer of `st.sortedSymbols` you assign the `make`. This may be reassigning a pointer which is safe or assigning to the content of the pointer. It's probably the former, but if the latter that is unsafe as you are essentially overriding the content of `oldSymbols` as well.
In any case since there is no sorted insertion here, i'm not sure why you do the whole reallocation, just append the items, simpler and quicker, and sometimes you will avoid reallocation if the slice has preallocated space.

Reply:
> Then to the pointer of `st.sortedSymbols` you assign the `make`. This may be reassigning a pointer which is safe or assigning to the content of the pointer.

It's definitely the former, I'm replacing the pointer stored in `st.sortedSymbols` with a new one, this doesn't affect the slice that is now in `oldSymbols`.

> In any case since there is no sorted insertion here, i'm not sure why you do the whole reallocation, just append the items, simpler and quicker, and sometimes you will avoid reallocation if the slice has preallocated space.

I'm reallocating to avoid an overallocation of the slice, which will never be released. Also regular insertion is likely to result in at least 1 reallocation anyway.